sdk/python/feast/sdk/utils/bq_util.py

```diff
@@ -0,0 +1,154 @@
+# Copyright 2018 The Feast Authors
+# 
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+# 
+#     https://www.apache.org/licenses/LICENSE-2.0
+# 
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import os
+import time
+
+import pandas as pd
+from google.cloud.bigquery.client import Client as BQClient
+from google.cloud.bigquery.job import ExtractJobConfig, DestinationFormat
+from google.cloud.bigquery.table import Table
+from google.cloud.storage import Client as GCSClient
+
+from feast.sdk.utils.gs_utils import is_gs_path, split_gs_path, gs_to_df
+
+
+def head(client, table, max_rows=10):
+    '''Get the head of the table. Retrieves rows from the given table at 
+        minimum cost
+    
+    Args:
+        client (google.cloud.bigquery.client.Client): bigquery client
+        table (google.cloud.bigquery.table.Table): bigquery table to get the 
+            head of
+        max_rows (int, optional): Defaults to 10. maximum number of rows to 
+            retrieve
+    
+    Returns:
+        pandas.DataFrame: dataframe containing the head of rows
+    '''
+
+    rows = client.list_rows(table, max_results=max_rows)
+    rows = [x for x in rows]
+    return pd.DataFrame(
+        data=[list(x.values()) for x in rows], columns=list(rows[0].keys()))
+
+
+def get_table_name(feature_id, storage_spec):
+    """
+    Get fully qualified BigQuery table name from a feature ID and its
+    storage spec
+    Args:
+        feature_id(str): ID of a feature
+        storage_spec(feast.specs.StorageSpec_pb2.StorageSpec): storage spec of
+        the feature
+
+    Returns:
+         str: fully qualified table name of the feature.
+
+    """
+    if "bigquery" != storage_spec.type:
+        raise ValueError("storage spec is not BigQuery storage spec")
+
+    try:
+        project = storage_spec.options["project"]
+        dataset = storage_spec.options["dataset"]
+    except KeyError:
+        raise ValueError("storage spec has empty project or dataset option")
+
+    table_name = "_".join(feature_id.split(".")[:2])
+    return ".".join([project, dataset, table_name])
+
+
+class TableDownloader:
+    def __init__(self):
+        self._bq = None
+        self._gcs = None
+
+    @property
+    def gcs(self):
+        if self._gcs is None:
+            self._gcs = GCSClient()
+        return self._gcs
+
+    @property
+    def bq(self):
+        if self._bq is None:
+            self._bq = BQClient()
+        return self._bq
+
+    def download_table_as_file(self, table_id, dest, staging_location,
+                               file_type):
```

Comment:
Is it possible for us to remove `staging_location` from the code? It really makes the API cumbersome if the user needs to provide the staging location every time. Can't this just be a Feast system staging location? That way the user doesnt have to deal with staged data.

Reply:
Or do we want to fix that later?

Reply:
Yes, let's fix this later, it's not so simple since the user needs to have permission to write to whatever the default staging location is.